WordPress/src/main/java/org/wordpress/android/ui/stockmedia/StockMediaPickerActivity.java

```diff
@@ -0,0 +1,490 @@
+        mThumbHeight = (int) (mThumbWidth * 0.75f);
+
+        Toolbar toolbar = findViewById(R.id.toolbar);
+        setSupportActionBar(toolbar);
+        ActionBar actionBar = getSupportActionBar();
+        if (actionBar != null) {
+            actionBar.setDisplayHomeAsUpEnabled(true);
+            actionBar.setDisplayShowTitleEnabled(true);
+        }
+        configureSearchView();
+
+        RecyclerView recycler = findViewById(R.id.recycler);
+        recycler.setLayoutManager(new GridLayoutManager(this, getColumnCount()));
+
+        mAdapter = new StockMediaAdapter();
+        recycler.setAdapter(mAdapter);
+
+        mSelectionBar = findViewById(R.id.container_selection_bar);
+        mTextInsert = findViewById(R.id.text_insert);
+        mTextInsert.setOnClickListener(new View.OnClickListener() {
+            @Override public void onClick(View v) {
+                uploadSelection();
+            }
+        });
+        findViewById(R.id.text_clear).setOnClickListener(new View.OnClickListener() {
+            @Override public void onClick(View v) {
+                mAdapter.clearSelection();
+            }
+        });
+
+        setupObserver();
+
+        if (savedInstanceState == null) {
+            mSite = (SiteModel) getIntent().getSerializableExtra(WordPress.SITE);
+            showEmptyView(true);
+        } else {
+            mSite = (SiteModel) savedInstanceState.getSerializable(WordPress.SITE);
+            mViewModel.readFromBundle(savedInstanceState);
+            if (savedInstanceState.containsKey(KEY_SELECTED_ITEMS)) {
+                ArrayList<Integer> selectedItems = savedInstanceState.getIntegerArrayList(KEY_SELECTED_ITEMS);
+                if (selectedItems != null) {
+                    mAdapter.setSelectedItems(selectedItems);
+                }
+            }
+        }
+    }
+
+    @Override
+    protected void onDestroy() {
+        if (mSearchView != null) {
+            mSearchView.setOnQueryTextListener(null);
+            mSearchView.setOnCloseListener(null);
+        }
+        super.onDestroy();
+    }
+
+    @Override
+    protected void onSaveInstanceState(Bundle outState) {
+        super.onSaveInstanceState(outState);
+        mViewModel.writeToBundle(outState);
+        if (mSite != null) {
+            outState.putSerializable(WordPress.SITE, mSite);
+        }
+        if (!mAdapter.mSelectedItems.isEmpty()) {
+            outState.putIntegerArrayList(KEY_SELECTED_ITEMS, mAdapter.mSelectedItems);
+        }
+    }
+
+    /*
+     * observe when the ViewModel's search results have changed and update the adapter with them
+     */
+    private void setupObserver() {
+        mViewModel.getSearchResults().observe(this, new Observer<List<StockMediaModel>>() {
+            @Override
+            public void onChanged(@Nullable final List<StockMediaModel> mediaList) {
+                if (!isFinishing()) {
+                    if (!mViewModel.isFetching()) {
```

Comment:
I think the idea with view models and observer pattern is that the a change to one model doesn't depend on another. I don't know if that makes sense at all, but I'll try to explain with an example. Let's say we did a fetch and the stock media list has not changed for some reason. If we were to handle the `isFetching` inside of the `onChanged` of the underlying model, we _have to_ say that the list of stock media changed even if it hasn't so that the view will update it's fetching status, hence stop the progress bar.

A better way might be to have a different live data instance just for the status of search. [Here](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/WordPress/src/main/java/org/wordpress/android/viewmodel/PluginBrowserViewModel.java#L71) is an example of this. Ideally, the activity shouldn't even start the progress bar, it should just observe that status and if it says it's fetching, it should show the spinner, if it says it's not fetching it should stop the progress bar. That should be all the activity know about. This gives us a way to add some custom checkers in the view model side of things, so for example if there is a request going on for the same query, maybe we don't start a fetch at all in which case the progress bar won't show up. Or, let's say for whatever reason, we start another fetch just from the view model (not a user action), for example when the user first opens the page, then we'll still easily show the progress bar.

Hope this all makes sense.